### DISCUSSION TITLE
can we mock EWS APIs in flask

### DISCUSSION THREAD

**Discussion by prashanttest-elastic** [ecederstrand/exchangelib#1216] [General]
I want to mock the exchange server. I have written a code to fetch data from the exchange server. Now I want to redirect APIs to my Flask server and return the mock data.

**Comment by ecederstrand** [comment]
The exchangelib test suite uses a couple of different techniques for mocking. There's an example of service-specific mocking in https://github.com/ecederstrand/exchangelib/blob/490f734b5ba73b4038ad6c78f13fb23572b00228/tests/test_protocol.py#L582 and more general HTTP mocking in https://github.com/ecederstrand/exchangelib/blob/490f734b5ba73b4038ad6c78f13fb23572b00228/tests/test_autodiscover.py#L249

**Reply by moxarth** [reply]
@ecederstrand would it be possible to mock EWS APIs using the Flask framework?

**Reply by moxarth** [reply]
Also, can you please provide me an endpoint used for fetching mails (also for contacts, tasks and calendars if possible) and an example of a response the APIs are returning? It would be better for us to mock those API endpoints.

**Reply by ecederstrand** [reply]
I'm not sure what you're attempting to do. If you created a full EWS API clone in Flask, then I wouldn't call it mocking but rather live server testing. Just spin up your Flask server and point exchangelib to that URL instead using `Configuration.service_endpoint`.

If you want to redirect specific API calls to Flask and the rest to a real Exchange server, then use `requests_mock.mock(real_http=True)` as described above, and call into your Flask app to inject the HTTP response as needed.

**Reply by ecederstrand** [reply]
If you want to see examples of actual responses from the Exchange server, then enable debug logging (see https://ecederstrand.github.io/exchangelib/#troubleshooting) and check out the XML requests and responses.
